elasticsearch/security_api.go

```diff
@@ -0,0 +1,220 @@
+	err := response.DecodeTo(&confirmation)
+	return confirmation, err
+}
+
+func HasPrivileges(client Client, privileges HasPrivilegesRequest, credentials string) (HasPrivilegesResponse, error) {
+	h := fmt.Sprintf("Authorization: ApiKey %s", credentials)
+	response := client.JSONRequest(http.MethodGet, "/_security/user/_has_privileges", privileges, h)
+
+	var info HasPrivilegesResponse
+	err := response.DecodeTo(&info)
+	return info, err
+}
+
+type CreateApiKeyRequest struct {
+	Name            string         `json:"name"`
+	Expiration      *string        `json:"expiration,omitempty"`
+	RoleDescriptors RoleDescriptor `json:"role_descriptors"`
+}
+
+type CreateApiKeyResponse struct {
+	ApiKey
+	Key string `json:"api_key"`
+}
+
+type GetApiKeyRequest struct {
+	ApiKeyQuery
+	Owner bool `json:"owner"`
+}
+
+type GetApiKeyResponse struct {
+	ApiKeys []ApiKeyResponse `json:"api_keys"`
+}
+
+type CreatePrivilegesRequest map[AppName]PrivilegeGroup
+
+type CreatePrivilegesResponse map[AppName]PrivilegeResponse
+
+type HasPrivilegesRequest struct {
+	// can't reuse the `Applications` type because here the JSON attribute must be singular
+	Applications []Application `json:"application"`
+}
+type HasPrivilegesResponse struct {
+	Username    string                            `json:"username"`
+	HasAll      bool                              `json:"has_all_requested"`
+	Application map[AppName]PrivilegesPerResource `json:"application"`
+}
+
+type InvalidateApiKeyRequest struct {
+	ApiKeyQuery
+}
+
+type InvalidateApiKeyResponse struct {
+	Invalidated []string `json:"invalidated_api_keys"`
+	ErrorCount  int      `json:"error_count"`
+}
+
+type DeletePrivilegeRequest struct {
+	Application AppName       `json:"application"`
+	Privilege   PrivilegeName `json:"privilege"`
+}
+
+//noinspection GoRedundantParens
+type DeletePrivilegeResponse map[AppName](map[PrivilegeName]DeleteResponse)
+
+type RoleDescriptor map[AppName]Applications
+
+type Applications struct {
+	Applications []Application `json:"applications"`
+}
+
+type Application struct {
+	Name       AppName     `json:"application"`
+	Privileges []Privilege `json:"privileges"`
+	Resources  []Resource  `json:"resources"`
+}
+
+type ApiKeyResponse struct {
+	ApiKey
+	Creation    int64  `json:"creation"`
+	Invalidated bool   `json:"invalidated"`
+	Username    string `json:"username"`
+}
+
+type ApiKeyQuery struct {
+	// normally the Elasticsearch API will require either Id or Name, but not both
+	Id   *string `json:"id,omitempty"`
+	Name *string `json:"name,omitempty"`
+}
+
+type ApiKey struct {
+	Id           string `json:"id"`
+	Name         string `json:"name"`
+	ExpirationMs *int64 `json:"expiration,omitempty"`
+	// This attribute does not come from Elasticsearch, but is filled in by APM Server
+	Credentials *string `json:"credentials,omitempty"`
+}
+
+type PrivilegeResponse map[Privilege]PutResponse
+
+type PrivilegeGroup map[PrivilegeName]Actions
+
+type Perms map[Privilege]bool
```

Comment:
Don't use arbitrary abbreviation here, but rather `Permissions` - that's what we agreed on.

Reply:
I got used to term enough to not consider it arbitrary, but: sure thing

https://github.com/elastic/apm-server/search?q=perms&type=Code